### PR TITLE
Align and synchronize restore behavior between SessionManager and BrowserStore

### PR DESCRIPTION
**Pull request by pocmo**
Issue #3557 and issue #3432.

### CODE REVIEW

File: components/browser/session/src/main/java/mozilla/components/browser/session/SessionManager.kt [hunk 1]
```diff
@@ -122,7 +122,32 @@ class SessionManager(
      * @param snapshot A [Snapshot] which may be produced by [createSnapshot].
      * @param updateSelection Whether the selected session should be updated from the restored snapshot.
      */
-    fun restore(snapshot: Snapshot, updateSelection: Boolean = true) = delegate.restore(snapshot, updateSelection)
+    fun restore(snapshot: Snapshot, updateSelection: Boolean = true) {
+        delegate.restore(snapshot, updateSelection)
+
+        val tabs = snapshot.sessions.mapNotNull { item ->
```
Comment (csadilek) [inline components/browser/session/src/main/java/mozilla/components/browser/session/SessionManager.kt]: nit: filter then map would be easier to read (but that's subjective) :)

File: components/browser/session/src/main/java/mozilla/components/browser/session/SessionManager.kt [hunk 2]
```diff
@@ -122,7 +122,32 @@ class SessionManager(
...
+            } else {
+                // A restored snapshot should not contain any custom tab so we should be able to safely ignore
+                // them here.
+                null
+            }
+        }
+
+        val selectedTabId = if (updateSelection && snapshot.selectedSessionIndex != NO_SELECTION) {
+            val index = snapshot.selectedSessionIndex
+            if (index in 0..snapshot.sessions.lastIndex) {
```
Comment (csadilek) [inline components/browser/session/src/main/java/mozilla/components/browser/session/SessionManager.kt]: nit: index <= lastIndex?

File: components/browser/session/src/test/java/mozilla/components/browser/session/SessionManagerMigrationTest.kt [hunk 3]
```diff
@@ -200,4 +200,100 @@ class SessionManagerMigrationTest {
         assertNull(manager.selectedSession)
         assertNull(store.state.selectedTabId)
     }
+
+    @Test
+    fun `Restoring snapshot with invalid index`() {
+        val store = BrowserStore(BrowserState())
```
Comment (csadilek) [inline components/browser/session/src/test/java/mozilla/components/browser/session/SessionManagerMigrationTest.kt]: nit: don't need to provide empty state.

File: components/browser/session/src/main/java/mozilla/components/browser/session/SessionManager.kt (hunk 2 above)
Comment (pocmo) [reply components/browser/session/src/main/java/mozilla/components/browser/session/SessionManager.kt]: Not sure about this comment. We also need to check >= 0 (at least no selection is -1).

Comment (csadilek) [reply components/browser/session/src/main/java/mozilla/components/browser/session/SessionManager.kt]: Hm not sure what I wanted to say there :).

>We also need to check >= 0 (at least no selection is -1).
Yeah, makes sense.

**Pull request merged**